File: swift/1283_FindTheSmallestDivisorGivenAThreshold.swift
class Solution {
    func smallestDivisor(_ nums: [Int], _ threshold: Int) -> Int {
        func computeSum(_ divisor: Int) -> Int {
            var total = 0
            for n in nums {
                total += n / divisor + (n % divisor != 0 ? 1 : 0)
            }
            return total            
        }
        
        var left = 1, right = 2
        while computeSum(right) > threshold {
            left = right
            right <<= 1
        }
        
        while left <= right {
            var pivot = left + (right - left) / 2
            var total = computeSum(pivot)
            if total > threshold {
                left = pivot + 1
            } else {
                right = pivot - 1
            }
        }
        return left
    }
}

File: python/1283_FindTheSmallestDivisorGivenAThreshold.py
import functools
class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        computeSum = lambda x : functools.reduce(lambda a, b : a + ceil(b / x), nums, 0)
        
        left, right = 1, 2
        while computeSum(right) > threshold:
            left, right = right, right << 1
        
        while left <= right:
            pivot = (left + right) >> 1
            total = computeSum(pivot)
            
            if total > threshold: left = pivot + 1
            else: right = pivot - 1
                
        return left
